File: math/happy_number.rb
# @param {Integer} n
# @return {Boolean}
def is_happy(n)
    seen = Set.new
    
    while n != 1 && !seen.include?(n)
        seen.add(n)

        sum = 0
        n.to_s.each_char do |digit|
            sum += digit.to_i ** 2
        end

        n = sum
    end

    return n == 1
end

File: math/happy_number.py
class Solution:
    def isHappy(self, n: int) -> bool:
        seen = set()

        while n != 1 and n not in seen:
            seen.add(n)

            total = sum(int(digit) ** 2 for digit in str(n))
            n = total

        return n == 1
    